src/@components/login/loginModal.tsx

```diff
@@ -0,0 +1,265 @@
+import styled, { css } from "styled-components";
+import {
+  ProducerDefaultModeToggleIc,
+  ProducerModeToggleIc,
+  EyeIc,
+  ProducerLoginBtnIc,
+  VocalLoginBtnIc,
+  DefaultLoginBtnIc,
+} from "../../assets";
+import { useState } from "react";
+import { Link } from "react-router-dom";
+
+export default function LoginModal() {
+  const [isProducerMode, setIsProducerMode] = useState<boolean>(false);
+  const [emailInputState, setEmailInputState] = useState<string>("");
+  const [passwordInputState, setPasswordInputState] = useState<string>("");
+  const [isPasswordType, setIsPasswordType] = useState<boolean>(true);
+  const [emailDefaultState, setEmailDefaultState] = useState<boolean>(true);
+  const [passwordDefaultState, setPasswordDefaultState] = useState<boolean>(true);
+
+  const EMAIL_RULE = /^[0-9a-zA-Z]([-_\.]?[0-9a-zA-Z])*@[0-9a-zA-Z]([-_\.]?[0-9a-zA-Z])*\.[a-zA-Z]{2,3}$/i;
+  const PASSWORD_RULE = /^(?=.*[a-zA-Z])((?=.*\d)(?=.*\W)).{8,16}$/;
+
+  const FOCUS = "focus";
+  const BLUR = "blur";
+  const WARNING = "warning";
+
+  function producerToggleType() {
+    return isProducerMode ? (
+      <ProducerModeToggleIc onClick={() => setIsProducerMode((prev) => !prev)} style={{ cursor: "pointer" }} />
+    ) : (
+      <ProducerDefaultModeToggleIc onClick={() => setIsProducerMode((prev) => !prev)} style={{ cursor: "pointer" }} />
+    );
+  }
+
+  function changeHoverState(
+    e: React.FocusEvent<HTMLInputElement>,
+    setInputState: React.Dispatch<React.SetStateAction<string>>,
+    RULE: RegExp,
+  ): void {
+    const input = e.target.value;
+
+    if (!RULE.test(input) && !isInputEmpty(input)) setInputState(WARNING);
+    if (RULE.test(input)) {
+      e.type === "focus" ? setInputState("focus") : setInputState(BLUR);
+    }
+    if (isInputEmpty(input)) {
+      e.type === FOCUS ? setInputState(FOCUS) : setInputState(BLUR);
+    }
```

Comment:
```suggestion
    if (!RULE.test(input) && !isInputEmpty(input)) {
    setInputState(WARNING);
    return
    }
      e.type === FOCUS ? setInputState(FOCUS) : setInputState(BLUR);
    }
```

오빠가 전에 말했던 부분 같은데 유효성 검사하는 부분은 확실히 빼는 게 좋을 것 같아요!!
정규표현식부분이 상수는 맞다고 생각하는데, 보통 정규표현식을 매직넘버로 빼는지는 잘 모르겟다! 근데 난 좋다고 생각해👍

이부분 수정이 필요한 것 같은데

위에 if문에 끝나면 밑에 실행 될 필요가 없네요! if문에 끝나고 바로 return을 해서 불필요한 조건문을 수행하지 않게 하는게 좋을 것 같아요. (early return)
RULE.test(input) 일때랑 isInputEmpty(input)일때랑 수행해야하는 로직이 같네요? 중복된 코드를 없앱시다!

Reply:
와... 웅니 짱이다

Reply:
💗💖💘💝